utils/SimpleJson.cpp

```diff
@@ -0,0 +1,410 @@
+// including " or }.
+const char* SimpleJson::findValueEnd(const char* start) const
+{
+    auto cursor = start;
+    if (!start)
+        return nullptr;
+
+    cursor = eatWhiteSpaces(cursor);
+    if (!cursor)
+        return nullptr;
+    switch (*cursor)
+    {
+    case '{':
+        return findObjectEnd(cursor);
+    case '[':
+        return findArrayEnd(cursor);
+    case '"':
+        return findStringEnd(cursor);
+    case 't':
+    case 'f':
+        return findBooleanEnd(cursor);
+    case 'n':
+        return findNullEnd(cursor);
+    }
+
+    return findNumberEnd(cursor);
+}
+
+// Finds the property on the current level.
+// start should point at the " or whitespace character.
+// Returns SimpleJson object of the property's value
+SimpleJson SimpleJson::findProperty(const char* start, const std::string& name) const
+{
+    auto cursor = start;
+    if (name.empty() || !start || ('"' != *start && !std::isspace(*start)))
+        return SimpleJsonNone;
+
+    while (cursor <= _cursorOut)
+    {
+        cursor = eatWhiteSpaces(cursor);
+        if (!cursor)
+            return SimpleJsonNone;
+        if ('"' != *cursor)
+            return SimpleJsonNone;
+        // Read property name.
+        auto propName = cursor;
+        auto propNameEnd = findStringEnd(cursor);
+        if (!propNameEnd)
+            return SimpleJsonNone;
+        cursor = eatWhiteSpaces(propNameEnd + 1);
+        if (!cursor)
+            return SimpleJsonNone;
+        if (':' != *cursor)
+            return SimpleJsonNone;
+        // Read property value.
+        auto valueStart = cursor + 1;
+        if (valueStart >= _cursorOut)
+            return SimpleJsonNone;
+        auto valueEnd = findValueEnd(cursor + 1);
+        if (!valueEnd)
+            return SimpleJsonNone;
+        // Check whether it is our property.
+        if (!strncmp(propName + 1, name.c_str(), name.length()))
+        {
+            return SimpleJson::create(valueStart, valueEnd);
+        }
+        // Go for the next one.
+        cursor = eatWhiteSpaces(valueEnd + 1);
+        if (!cursor)
+            return SimpleJsonNone;
+        if (',' != *cursor)
+            return SimpleJsonNone;
+        cursor++;
+    }
+
+    return SimpleJsonNone;
+}
+
+SimpleJson SimpleJson::find(const std::string& path)
+{
+    auto node = _nodeCache.find(path);
+    if (node != _nodeCache.end())
+    {
+        return node->second;
+    }
+    std::string cachedPath = "";
+    return findInternal(path, cachedPath, _nodeCache);
+}
+
+SimpleJson SimpleJson::findInternal(const std::string& path,
+    std::string& cachedPath,
+    std::map<std::string, SimpleJson>& nodeCache)
+{
+    auto token = StringTokenizer::tokenize(path.c_str(), path.length(), '.');
+    if (token.empty() || Type::Object != _type)
+        return SimpleJsonNone;
+
+    if (!cachedPath.empty())
+    {
+        cachedPath += '.';
+    }
+    cachedPath += token.str();
```

Comment:
this is building with new allocations

Reply:
Fixed. All std::string operations removed. CachedPath is constructed in a fixed size buffer, which should be enough for our applications.